src/audio/codec_adapter/codec_adapter.c

```diff
@@ -404,19 +396,21 @@ int codec_adapter_copy(struct comp_dev *dev)
 		goto db_verify;
 	}
 
-	/* Process only we have enough free data in the local
-	 * buffer. If we don't have enough free space process()
-	 * will override the data in the local buffer
-	 */
-	if (local_buff->stream.free < get_output_bytes(dev))
-		goto db_verify;
-
 	buffer_stream_invalidate(source, codec_buff_size);
 	codec_adapter_copy_from_source_to_lib(&source->stream, &codec->cpd,
 					      codec_buff_size);
 	codec->cpd.avail = codec_buff_size;
 	ret = codec_process(dev);
 	if (ret) {
+		if (ret == -ENOSPC) {
+			/* copy the data back to the source buffer as it was not used */
+			codec_adapter_copy_from_lib_to_sink(&codec->cpd, &source->stream,
+							    codec_buff_size);
+			buffer_stream_writeback(source, codec_buff_size);
+			codec->cpd.avail -= codec_buff_size;
+			goto db_verify;
```

Comment:
@ranj063 i figured out what's the problem. You do not need to copy back the data and to decrease cpd.avail because the data is not already consumed.

So, a thing like this:

```
        codec->cpd.avail = codec_buff_size;
        ret = codec_process(dev);
        if (ret) {
+               if (ret == -ENOSPC) {
+                       ret = 0;
+                       goto db_verify;
+               }
+
```
makes everything working.

Reply:
Thank you so much @dbaluta. Updated now!